crates/cadenza-eval/src/eval.rs

```diff
@@ -0,0 +1,458 @@
+//! Tree-walk evaluator for the Cadenza language.
+//!
+//! The evaluator interprets the AST produced by cadenza-syntax.
+//! It handles:
+//! - Literals (integers, floats, strings)
+//! - Identifiers (variable lookup)
+//! - Lists/vectors
+//! - Function application
+//! - Macro expansion
+
+use crate::compiler::Compiler;
+use crate::env::Env;
+use crate::error::{Error, Result};
+use crate::interner::{InternedId, Interner};
+use crate::value::Value;
+use cadenza_syntax::ast::{Apply, Attr, Expr, Ident, Literal, LiteralValue, Root, Synthetic};
+
+/// Evaluates a complete source file (Root node).
+///
+/// Each top-level expression is evaluated in order. The results are
+/// collected into a vector, though most top-level expressions will
+/// return `Value::Nil` as side effects on the `Compiler` are the
+/// primary purpose.
+pub fn eval(root: &Root, interner: &mut Interner, env: &mut Env, compiler: &mut Compiler) -> Result<Vec<Value>> {
+    let mut results = Vec::new();
+    for expr in root.items() {
+        let value = eval_expr(&expr, interner, env, compiler)?;
```

Comment:
I would like to avoid bailing at the top-level. Errors should be stored in the compiler so we can have more than one. And then they can be attached to syntax and emitted either in the cli or LSP server.